plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,1001 @@
+						FuChunk *chk,
+						GError **error)
+{
+	g_autoptr(GByteArray) cod_ff = g_byte_array_new();
+	const guint32 pkt_size = fu_chunk_get_data_sz(chk);
+	const guint8 buf[2] = {DDC_OPCODE_MTK_REQ, DDC_VCP_CODE_SET_DATA_FF};
+	guint8 data[4] = {0x00};
+
+	if (!fu_memcpy_safe(data,
+			    sizeof(data),
+			    0x0,
+			    (const guint8 *)&pkt_size,
+			    sizeof(pkt_size),
+			    0x0,
+			    sizeof(pkt_size),
+			    error))
+		return FALSE;
+	g_byte_array_append(cod_ff, buf, sizeof(buf));
+	g_byte_array_append(cod_ff, data, sizeof(data));
+	return mediatek_scaler_device_ddc_write(self, cod_ff, error);
+}
+
+static gboolean
+fu_mediatek_scaler_device_write_firmware_impl(FuMediatekScalerDevice *self,
+					      GBytes *fw,
+					      FuProgress *progress,
+					      GError **error)
+{
+	g_autoptr(GPtrArray) chunks = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	chunks = fu_chunk_array_new_from_bytes(fw, 0x00, 0x00, DDC_DATA_PAGE_SIZE);
+	for (gint retry = 1; retry <= DDC_RW_MAX_RETRY_CNT; retry++) {
+		g_clear_error(&error_local);
+
+		for (guint i = 0; i < chunks->len; i++) {
+			FuChunk *chk = g_ptr_array_index(chunks, i);
+
+			/* fast forward if chunk is empty, on the contrary set data per fragment
+			 * size */
+			if (fu_mediatek_scaler_device_chunk_data_is_blank(chk)) {
+				if (!fu_mediatek_scaler_device_set_data_fast_forward(self,
+										     chk,
+										     error))
+					return FALSE;
+			} else {
+				if (!fu_mediatek_scaler_device_set_data(self, chk, error))
+					return FALSE;
+				fu_device_sleep(FU_DEVICE(self), 1);
+			}
+
+			/* update progress */
+			fu_progress_set_percentage_full(fu_progress_get_child(progress),
+							(gsize)i + 1,
+							(gsize)chunks->len);
+		}
+
+		/* exit the try loop when successes */
+		fu_device_sleep(FU_DEVICE(self), FU_MEDIATEK_SCALER_DDC_MSG_DELAY_MS);
+		if (fu_mediatek_scaler_device_check_sent_info(self, fw, &error_local))
+			return TRUE;
+
+		if (!g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_WRITE)) {
+			g_propagate_error(error, g_steal_pointer(&error_local));
+			return FALSE;
+		}
+
+		g_debug("retry write_firmware: step: %d, max: %d", retry, DDC_RW_MAX_RETRY_CNT);
+	}
+	g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "maximum tries exceeded");
+	return FALSE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	guint max_tries = 30;
+
+	/* open device */
+	locker = fu_device_locker_new(self->i2c_dev, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* set the target address */
+	if (!fu_mediatek_scaler_ensure_device_address(self, DDC_DISPLAY_I2C_ADDR >> 1, error))
+		return FALSE;
+
+	/* reboot the device */
+	if (!(fu_mediatek_scaler_device_set_isp_reboot(self, error)))
+		return FALSE;
+
+	/* wait for the device back */
+	if (!fu_device_retry_full(device,
```

Comment:
Doest it make sense to `retry` last until the device is back from rebooting? or a delay time instead?